File: labs/capstone-4-agent-team/domain-b-ecommerce/solution/coordinator.js
/**
 * Pipeline Coordinator — B2B Ecommerce Order Pipeline (Node.js Solution)
 */

import Anthropic from "@anthropic-ai/sdk";

const MODEL = "claude-sonnet-4-20250514";
const MAX_ITERATIONS = 10;

const ORDERS = {
  "PO-2024-5001": {
    order_id: "PO-2024-5001",
    customer_name: "Apex Manufacturing Corp",
    customer_id: "CUST-100",
    items: [
      { sku: "WDG-4420", qty: 500, unit_price: 12.5 },
      { sku: "BLT-7780", qty: 2000, unit_price: 0.85 },
    ],
    sla_tier: "standard",
    requested_delivery: "2024-12-01",
  },
};

const INVENTORY = {
  "WH-EAST": { "WDG-4420": 8000, "BLT-7780": 50000 },
  "WH-CENTRAL": { "WDG-4420": 12000, "BLT-7780": 30000 },
  "WH-WEST": { "WDG-4420": 5000, "BLT-7780": 20000 },
};

function validateOrder(orderId) {
  const order = ORDERS[orderId];
  if (!order) return { valid: false, errors: ["Not found"] };
  const errors = [];
  for (const item of order.items) {
    let found = false;
    for (const wh of Object.values(INVENTORY)) {
      if (item.sku in wh) found = true;
    }
    if (!found) errors.push(`Unknown SKU: ${item.sku}`);
    if (item.qty <= 0) errors.push(`Invalid qty: ${item.sku}`);
  }
  return { valid: errors.length === 0, errors };
}

function checkInventory(orderId) {
  const order = ORDERS[orderId];
  if (!order) return { error: "Not found" };
  const result = {};
  for (const item of order.items) {
    const byWh = {};
    let total = 0;
    for (const [whId, stock] of Object.entries(INVENTORY)) {
      const qty = stock[item.sku] || 0;
      byWh[whId] = qty;
      total += qty;
    }
    result[item.sku] = { by_warehouse: byWh, total, needed: item.qty, sufficient: total >= item.qty };
  }
  return result;
}

class CircuitBreaker {
  constructor(maxConsec = 3) {
    this.maxConsec = maxConsec;
    this.consec = 0;
    this.state = "closed";
  }
  recordSuccess() { this.consec = 0; if (this.state === "half_open") this.state = "closed"; }
  recordFailure() { this.consec++; if (this.consec > this.maxConsec) this.state = "open"; }
  isTripped() { return this.state === "open"; }
  reset() { this.consec = 0; this.state = "closed"; }
}

async function runPipeline(orderId) {
  const order = ORDERS[orderId];
  if (!order) { console.log(`Order ${orderId} not found`); return; }
  console.log(`\n${"#".repeat(50)}\nPipeline for ${orderId} -- ${order.customer_name}\n${"#".repeat(50)}`);

  const v = validateOrder(orderId);
  console.log(`[Intake] Validation: ${v.valid ? "PASS" : "FAIL"}`);

  const inv = checkInventory(orderId);
  console.log(`[Intake] Inventory checked`);

  // Simplified fulfillment
  console.log(`[Fulfillment] Allocating from WH-CENTRAL`);
  console.log(`[Fulfillment] Carrier: ExpressLine Logistics`);

  // Exception check
  console.log(`[Monitor] SLA: on_track`);

  // Communication
  console.log(`[Comm] Customer update sent`);
  console.log(`[Comm] Event logged`);

  console.log(`\n${"*".repeat(50)}\nPipeline COMPLETE\n${"*".repeat(50)}`);
}

async function main() {
  await runPipeline("PO-2024-5001");
}

main().catch(console.error);

File: labs/capstone-4-agent-team/domain-b-ecommerce/solution/coordinator.py
"""
Pipeline Coordinator — B2B Ecommerce Order Pipeline (Solution)
"""

import json
import uuid
from datetime import datetime
from agents import PipelineState
from agents.agent1 import OrderIntakeAgent
from agents.agent2 import FulfillmentPlanningAgent
from agents.agent3 import ExceptionMonitorAgent
from agents.agent4 import CommunicationAgent
from quality_gate import CircuitBreaker
from mock_data import ORDERS

_circuit_breaker = CircuitBreaker(name="sla_violations", max_consecutive_failures=3)


def log_transition(pid, from_a, to_a, state):
    print(f"\n{'='*60}")
    print(f"[{datetime.now().isoformat()}] {pid}: {from_a} -> {to_a} (halted={state.halted})")
    print(f"{'='*60}")


def run_pipeline(order_id):
    order = ORDERS.get(order_id)
    if not order:
        return PipelineState(halted=True, halt_reason=f"Order {order_id} not found")

    pid = f"PL-{uuid.uuid4().hex[:8].upper()}"
    state = PipelineState(pipeline_id=pid, started_at=datetime.now().isoformat(), raw_order=order)
    print(f"\n{'#'*60}\n# Pipeline {pid} -- {order_id} ({order.get('customer_name')})\n{'#'*60}")

    if _circuit_breaker.is_tripped():
        state.halted = True
        state.halt_reason = "Circuit breaker OPEN"
        return state

    # Agent 1
    state.current_agent = "OrderIntakeAgent"
    log_transition(pid, "START", "OrderIntakeAgent", state)
    state = OrderIntakeAgent().run(state)
    if not state.intake.validation_passed:
        _circuit_breaker.record_failure()
    else:
        _circuit_breaker.record_success()
    if state.halted:
        return state

    # Agent 2
    state.current_agent = "FulfillmentPlanningAgent"
    log_transition(pid, "OrderIntakeAgent", "FulfillmentPlanningAgent", state)
    state = FulfillmentPlanningAgent().run(state)
    if state.halted:
        return state

    # Agent 3
    state.current_agent = "ExceptionMonitorAgent"
    log_transition(pid, "FulfillmentPlanningAgent", "ExceptionMonitorAgent", state)
    state = ExceptionMonitorAgent().run(state)
    if state.exception.sla_status == "violated":
        _circuit_breaker.record_failure()
    if _circuit_breaker.is_tripped():
        state.halted = True
        state.halt_reason = "Circuit breaker tripped: consecutive SLA violations"
        return state
    if state.halted:
        return state

    # Agent 4
    state.current_agent = "CommunicationAgent"
    log_transition(pid, "ExceptionMonitorAgent", "CommunicationAgent", state)
    state = CommunicationAgent().run(state)
    state.completed = True
    log_transition(pid, "CommunicationAgent", "COMPLETED", state)

    print(f"\n{'*'*60}\nPipeline {pid} COMPLETE\n  Order: {order_id}\n  SLA: {state.exception.sla_status}\n  Carrier: {state.fulfillment.selected_carrier}\n{'*'*60}")
    return state


def main():
    _circuit_breaker.reset()
    print("=== TEST 1: Happy Path ===")
    run_pipeline("PO-2024-5001")

    print("\n=== TEST 2: Split Shipment (HITL) ===")
    run_pipeline("PO-2024-5007")

    print("\n=== TEST 3: Invalid SKU ===")
    run_pipeline("PO-2024-5006")


if __name__ == "__main__":
    main()
